Add easy setup (installation of TRAL and external software)

LinaNaef opened the pull request:
added scripts for easy installation and external software

LinaNaef committed:
This branch contains a draft for new TRAL setup.
- automated setup of TRAL
	- automated creation of path system
	- configTRAL.cfg to set path
- automated download and installation of external software for TRAL
	- external software can only be installed one by one (licencing)
	- only for ALF and HMMER reliable installation in this commit
	- other external software only with draft .sh files
- automated deletion of TRAL and external software
- .sh file to activate TRAL and set PATH variables

LinaNaef committed:
- changed TRAL_EXT and TRAL_SOFTWARE to one single directory TRAL_EXT_SOFTWARE
	- adapted these changes in all files
- adapted deleteTRAL.sh to new requirements (draft)
- some small adaptations in all files

LinaNaef committed:
- change in installation path for external software
	ext software will be installed either in there default installing path or put in usr/local/bin
- outcommented uninstallation proceture after all .sh files done till now
- new mafft.sh
- new phobos.sh

LinaNaef committed:
- all external software will be installed in its default installing path or usr/local/bin
- uninstallation (outcommented) within all .sh files + uninstall_all_ext_software.sh
- folder alternative_installation for scripts which uses not default way to install external software
- tredparse.sh (instead of tred.sh) to install tredparse
- deleted all sudo within scripts
    - scripts has to be executed as root if they work within system path
- added installation of TRAL with PIP (ATTENTION: this version is not up to date!)

LinaNaef committed:
- integration of installation files for remaining external software
    - trust.sh
    - trf.sh
    - treks.sh (their website down, automated download failes till now)
- integrated option uninstall_all_ext_software.sh into deleteTRAL.sh
- changed in all scripts copy without overwriting (cp -n) to copy with overwrite (cp)
    reason: update files in systempath after change in script or in paths
- setupTRAL.sh: integrated function to install all external software which you like
- included test to see wheter a ext software already installed
- added some comments and structure to some scripts
- created first README for easy_setup

LinaNaef committed:
Merge pull request #4 from LinaNaef/setupTRAL

Easy_Setup for TRAL

LinaNaef committed:
T-Reks website adaption
- changed link
- activated link
- added --no-check-certificate to wget to prevent security error

LinaNaef committed:
Merge pull request #5 from LinaNaef/setupTRAL

T-Reks website adaption

LinaNaef committed:
Inclution of small comments and licenses within easy_setup ext_software scripts
	- for all external software a note regarding the license was included (if any was available)
	- added for all external software a small description
	- correction of some small typos

LinaNaef committed:
Merge branch 'develop' of https://github.com/LinaNaef/tral into develop

LinaNaef committed:
changed name of README to README.md

LinaNaef committed:
Update README.md

LinaNaef committed:
Update README.md

LinaNaef committed:
Merge branch 'develop' of https://github.com/acg-team/tral into develop

LinaNaef committed:
Individual script to install all external software and some bugfixes in easy_tral
- new script install_ext_software.sh
- bugfix in deleteTRAL.sh
- bugfix in mafft (added an absolute path)
- mafft.sh: test if installation package already downloaded
- changed cd in tredparse.sh to absolut path (bugfix: produced error in subsequent scripts: "No such file or directory")
- some correction of typos and some descriptions

LinaNaef committed:
Added HHrepID
- created file hhrepid.sh to install hhrepid_64 as external software
- adapted README for hhrepid
- adapted uninstall_ext_software.sh for hhrepid
- adapted install_ext_software.sh for hhrepid

LinaNaef committed:
corrected typo

LinaNaef committed:
Addition of XSTREAM
- created file xstream.sh to install xstream.sh as external software
- adapted README for xstream.sh
- adapted deleteTRAL.sh for xstream.sh
- adapted install_ext_software for xstream
- changed some small typos
- changed some names to access external software
- edited names in configuration file

LinaNaef committed:
Merge branch 'xstream' into develop
There I mainly added xstream to the internal software and corrected some typos.

sbliven committed:
Fix easy_setup to install under Ubuntu Bionic

- Allow python3 binary to be set in config (PYTHON3 and PIP3 variables)
- Remove references to python3.5
- Add some error handling (not comprehensive)

LinaNaef committed:
Merge pull request #6 from sbliven/add_easy_setup

Fix easy_setup to install under Ubuntu Bionic

LinaNaef committed:
adapt .gitignore to jupyter notebook

LinaNaef committed:
Merge branch 'add_easy_setup' of https://github.com/LinaNaef/tral into add_easy_setup

LinaNaef committed:
add jupyter to gitignore

LinaNaef committed:
parsing ext software & TRAL download test
- installation of all ext software with parsing through folder rather than explicit statement
- test if TRAL repository is already downloaded
- change argument in setupTRAL.sh from "git" to "setup"

LinaNaef committed:
Change installation paths
- set variable $INSTALLATION_PATH to usr/local/bin (can now be modified in configTRAL_path.cfg
- file system will now be set up within $FILES (default: $HOME)
- adapted some things in README.md
- fixed test if tral_repository exists

LinaNaef committed:
Error handling easy_setup
- included error catching in easy_setup
- added double quotes to paths to prevent path splits
- setupTRAL.sh: bugfix: cd into directory with setup.py (otherwise FileNotFoundError while Installing TRAL

LinaNaef committed:
Added ACCEPT_ALL option
- in case ACCEPT_ALL=yes tral will fully installed without asking for options

LinaNaef committed:
deleted unintended download file

LinaNaef committed:
Bugfix install tral with pip

sbliven commented on .gitignore:
```diff
@@ -54,6 +54,13 @@ docs/_build/
+# Jupyter Notebook
+*.ipynb
+
+<<<<<<< HEAD
```
It looks like a merge conflict wasn't fully resolved here
```suggestion
```

sbliven commented on easy_setup/configTRAL_path.cfg:
```diff
@@ -0,0 +1,30 @@
+
+PYTHON3=python3
+PIP3=pip
+ACCEPT_ALL=no                     # set to "yes" if you want to accept all reads through installation
```
```suggestion
ACCEPT_ALL=no                     # set to "yes" to accept all licenses and install all external software
```

sbliven replied:
It's good this defaults to 'no'. We might even want to remove it from the default config to make it harder to ignore licenses. What do you think?

LinaNaef replied:
You're right. This part is just for our debugging process, so people don't have to see it.

sbliven commented on easy_setup/configTRAL_path.cfg:
```diff
@@ -0,0 +1,30 @@
+TRAL_ENV=$TRAL_PATH/tral_env        # virtual environment for python3
+
+# Configuration files are added to home per default. Do not change here.
+# If you want to change this default behaviour, this had be done within setup.py of TRAL.
```
Actually, you can pass `--tral-home` to setup.py during installation to change this.

LinaNaef replied:
change this later

sbliven commented:
#22 was a duplicate of this PR, and mentions the following features which were added to this branch after opening:

- ErrorHandling
- ACCEPT_ALL option
- parse all bash scripts for external software
- change of installation path for TRAL and external software
- double quotes for all paths
- test if TRAL repository already downloaded

sbliven commented on easy_setup/setupTRAL.sh:
```diff
@@ -0,0 +1,199 @@
+while [ ! -x $(which virtualenv 2>/dev/null) ]; do
+    echo "Installing virtualenv required by: "${PIP:-pip}" install virtualenv."
+
+    if [[ "$ACCEPT_ALL" = "yes" ]] || [[ "$ACCEPT_ALL" = "Yes" ]]; then
```
These tests should be case insensitive. Either

```suggestion
if [[ "$ACCEPT_ALL" =~ ^[Yy]([Ee][Ss])?$ ]]; then
```

or, could use `shopt -s nocasematch` for the whole script.

sbliven commented on easy_setup/deleteTRAL.sh:
```diff
@@ -0,0 +1,91 @@
+### Delete TRAL
+
+# delete tral path system
+if [[ "$ACCEPT_ALL" = "yes" ]] || [[ "$ACCEPT_ALL" = "Yes" ]]; then
```
Should we use ACCEPT_ALL in the delete script? This is less likely to be automated. It could lead to very bad situations (e.g. if they had set TRAL_PATH=$HOME).

LinaNaef replied:
Thank you, very good point. I will remove it.

sbliven commented on easy_setup/setupTRAL.sh:
```diff
@@ -0,0 +1,199 @@
+    
+    
+    
+    elif [[ $1 == "pip" ]]; then
```
```suggestion
elif [[ $1 == "pip" ]]; then
```

sbliven commented on easy_setup/setupTRAL.sh:
```diff
@@ -0,0 +1,199 @@
+        exit 1
+        }
+        
+        } && {
```
Fix the indentation here and just above.

```suggestion
    } && {
```

sbliven commented on easy_setup/setupTRAL.sh:
```diff
@@ -0,0 +1,199 @@
+            }
+        ;;
+        [Nn]* )
+            echo -e "\nYou can download this files later from ftp: //ftp.vital-it.ch/papers/vital-it/Bioinformatics-Schaper.\n"
```
```suggestion
            echo -e "\nYou can download this file later from ftp://ftp.vital-it.ch/papers/vital-it/Bioinformatics-Schaper/pvalue.tar.gz and expand into $TRAL_CONF/data.\n"
```

sbliven commented on easy_setup/setupTRAL.sh:
```diff
@@ -0,0 +1,199 @@
+
+######################
+### Download p-Value distribution files
+
```
Should we also check that $TRAL_CONF was created during tral installation?

sbliven commented on easy_setup/uninstall_all_ext_software.sh:
```diff
@@ -0,0 +1,77 @@
+
+### ALF
+
+rm -rf /usr/local/bin/darwin*
```
better to list all options rather than using `*`

sbliven commented on tral/tral_configuration/config.ini:
```diff
@@ -8,14 +8,14 @@ sequence_type = AA
     [[repeat_detector_path]]
         # If the executable is in the system path, supply its name. Otherwise, supply the full path to the executable. Details are explained in TRAL's online docs.
-        PHOBOS = phobos
+        PHOBOS = phobos_*
```
Did you test this? I'd be surprised if the glob * got expanded.

LinaNaef replied:
I thought it worked but now I changed it to a symlink.

sbliven commented on tral/tral_configuration/config.ini:
```diff
@@ -8,14 +8,14 @@ sequence_type = AA
         T-REKS = T-REKS
         TRED = tred
-        TRF = trf
+        TRF = trf409.linux64
```
Instead, let's add a symlink in the TRF install script like `ln -s trf409.linux64 /usr/local/bin/trf` so the name is standard. This could be done for phobos too.

LinaNaef replied:
Thank you, I've integrated it.

sbliven commented on easy_setup/install_ext_software.sh:
```diff
@@ -0,0 +1,53 @@
+
+
+install_ext_software () {
+    
```
It would be good to print the license terms for each external software before accepting. That could go in a later PR though.

LinaNaef replied:
I put it into my ToDoes.

sbliven commented on easy_setup/install_ext_software/tredparse.sh:
```diff
@@ -0,0 +1,85 @@
+
+
+# install virtual environment called "tral2" with python2
+virtualenv "$TRAL_ENV/python2" -p python2
```
Use a `${PYTHON2:-python2}` so it can be overridden in the config file

sbliven reviewed (changes requested):
This is great, @LinaNaef! This will be very helpful. I've reviewed the code and marked a few minor things that you should change, but overall it looks good.

sbliven commented on easy_setup/install_ext_software.sh:
```diff
@@ -0,0 +1,53 @@
+case $yn in
+    [Yy]* )
+        echo -e "\n"
+        for software in "install_ext_software"/*.sh ; do install_ext_software $software ; done
```
You need to add error handling to the install_ext_software call.

sbliven committed:
Update .gitignore

Co-Authored-By: LinaNaef <37146750+LinaNaef@users.noreply.github.com>

LinaNaef committed:
Modifications in response to code-review #21
- resolve merge conflict .gitignore
- delete ACCEPT_ALL option in config
- delete ACCEPT_ALL option in deleteTRAL.sh
- added "shopt -s nocasematch" to make comparisons case-insensitive
- delete alternative_install
- check for $TRAL_CONF
- fix installation bug with phobos
- add symlink trf and phobos
- used ${PYTHON2:-python2}
- error handling to install_ext_software call

LinaNaef committed:
Merge branch 'add_easy_setup' of https://github.com/LinaNaef/tral into add_easy_setup

sbliven commented:
Looks good! I've tested this with vagrant and it installs without issues.

Let's merge, and then the last two unresolved issues can be fixed afterwards.
